Add latest sit script

robin-k-wilson opened the pull request:
**Test Plan for Sit Scripts:**

Setup:

1. Create sittable entity, can either:
     - Import from json: [Download Blue Chair JSON](https://hifi-production.s3.amazonaws.com/robin/sit/sitPrototypes/bluechairlatest.json)
     - Or add to an entity. Use the registration point to be the seat center. Ex - a stool would use a registration point of { x: 0.5, y: 1.0, z: 0.5 }
2. Ensure both the entity script and server script is running on Entity
3. Spawn multiple chairs. Best tested with multiple avatars and users see "Multiple chair/user" test section

**Desktop regular use case**
1. Walk within 3 m from the chair
2. “Click to Sit” overlay appears on chair flat on the seat and in the chair direction
3. Should fade to a barely visible dark grey over time
4. Click on the overlay
5. Pop into chair facing forward (without HMD overlay) and sit
6. Can swivel around in chair with mouse
7. Use any movement key and a “Hold to Stand” overlay will appear in front of view
8. Should pop in front of the chair facing the direction the avatar was facing

**HMD regular use case - differences from Desktop in bold**
1. Walk within 3 m from the chair
2. **“Trigger to Sit”** overlay appears on chair flat on the seat and in the chair direction
3. Should fade to a barely visible dark grey over time
4. Trigger on the overlay
5. **Overlay to Face Forward appears for one second for user to get comfortable in real space**
6. **Timed at the end of the overlay, Your avatar head and camera should be facing forward in the chair with your legs facing forward**
7. Can **snap** and rotate around in chair with **right joystick**
8. **Head can look around normally without popping out of the chair**
9. Use any movement key and a “Hold to Stand” overlay will appear in front of view
10. Should pop in front of the chair facing the direction the avatar was facing

**Overlay tests**
- Sittable overlay appears within 2.8 m of walking to the chair disappears after walking outside 5 m
- Face Forward and Hold to Stand overlay should stay in middle of view
- In HMD mode always see Trigger to Sit, in Desktop mode always see Click to Sit

**Get out of chair tests**
- In HMD, can move away from old position and pop out of chair
- Hold down movement keys including flying
- Disconnects from server then reloads it
- Pops out of chair when:
   - Resize avatar
   - Change Avatar
   - Change domain
   - Teleport out of chair

**Multiple chair/user tests**
- Overlay 3m/5m apply
- While sitting, can switch chairs using the overlay. Old chair become available for another person
- Click/trigger to sit overlay disappears shortly after another HMD user triggers to sit on chair

**Outlier cases to ensure do not happen**
- Avatar never has the sit animation active outside of a chair 
- 2 avatars can not sit in the same chair _- see Known Bug section_
- Should not see sittable overlay when another avatar is sitting in chair
- Chair does not lock up. If avatar disconnects suddenly, the heartbeat interval on the server script should allow chair to be settable again within about 11 seconds.

**Known Bugs - that happen <5% of the time**
- In HMD mode, after sit, avatar sometimes is not rotated forward in the chair
- 2 avatars sit in the same chair if they trigger/click at the exact same time
- Significant lag in a domain can cause an avatar in HMD to pop out of chair without sitting - can correct by trying again

hyperlogic commented:
@robin-k-wilson Please run eslint on this script, we want to ensure that this follows our standards.  Here is the list of eslint errors that need to be resolved:

```
e:\code\hifi\examples\robin-sit.js:19:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:20:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:27:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:29:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:31:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:33:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:34:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:35:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:36:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:38:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:40:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:41:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:43:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:44:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:45:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:46:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:47:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:48:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:50:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:51:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:53:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:54:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:56:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:57:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:59:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:60:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:60:9: 'SITTING_SEARCH_RADIUS' is assigned a value but never used. [Error/no-unused-vars]
e:\code\hifi\examples\robin-sit.js:62:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:64:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:65:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:66:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:68:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:77:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:78:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:79:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:80:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:83:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:84:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:85:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:86:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:87:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:88:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:89:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:91:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:92:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:93:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:94:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:94:129: Missing semicolon. [Error/semi]
e:\code\hifi\examples\robin-sit.js:95:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:97:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:98:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:99:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:100:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:102:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:104:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:105:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:107:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:109:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:170:14: Unexpected trailing comma. [Error/comma-dangle]
e:\code\hifi\examples\robin-sit.js:241:124: Unexpected trailing comma. [Error/comma-dangle]
e:\code\hifi\examples\robin-sit.js:242:22: Missing semicolon. [Error/semi]
e:\code\hifi\examples\robin-sit.js:254:126: Unexpected trailing comma. [Error/comma-dangle]
e:\code\hifi\examples\robin-sit.js:257:14: Unexpected trailing comma. [Error/comma-dangle]
e:\code\hifi\examples\robin-sit.js:313:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:343:13: Expected { after 'if' condition. [Error/curly]
e:\code\hifi\examples\robin-sit.js:349:13: Expected { after 'if' condition. [Error/curly]
e:\code\hifi\examples\robin-sit.js:366:61: 'HALF' is not defined. [Error/no-undef]
e:\code\hifi\examples\robin-sit.js:377:1: Line 377 exceeds the maximum line length of 128. [Error/max-len]
e:\code\hifi\examples\robin-sit.js:383:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:385:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:392:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:407:9: Expected { after 'if' condition. [Error/curly]
e:\code\hifi\examples\robin-sit.js:408:9: Expected { after 'if' condition. [Error/curly]
e:\code\hifi\examples\robin-sit.js:409:9: Expected { after 'if' condition. [Error/curly]
e:\code\hifi\examples\robin-sit.js:413:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:418:15: No magic number: 4. [Error/no-magic-numbers]
e:\code\hifi\examples\robin-sit.js:422:15: No magic number: 5. [Error/no-magic-numbers]
e:\code\hifi\examples\robin-sit.js:446:19: No magic number: 6. [Error/no-magic-numbers]
e:\code\hifi\examples\robin-sit.js:470:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:506:6: Unnecessary semicolon. [Error/no-extra-semi]
e:\code\hifi\examples\robin-sit.js:508:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:522:12: No magic number: 50. [Error/no-magic-numbers]
e:\code\hifi\examples\robin-sit.js:560:6: Unnecessary semicolon. [Error/no-extra-semi]
e:\code\hifi\examples\robin-sit.js:563:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:576:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:595:71: Expected '!==' and instead saw '!='. [Error/eqeqeq]
e:\code\hifi\examples\robin-sit.js:677:6: Unnecessary semicolon. [Error/no-extra-semi]
e:\code\hifi\examples\robin-sit.js:679:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:682:5: Expected indentation of 5 spaces but found 4. [Error/indent]
e:\code\hifi\examples\robin-sit.js:693:1: Expected indentation of 1 space but found 0. [Error/indent]

90 problems
```

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+                var changeAlpha = 0.01;
+                overlayIntervalTransparency = Script.setInterval(function () {
+                    startAlpha = startAlpha - changeAlpha; // My new alpha
+                    Overlays.editOverlay(overlaySittable, { alpha: startAlpha }); // Edit the existing overlay
```
you should probably check that overlaSittable is not null, before calling editOveraly.

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+                    str = i + 1;
+                    overlayPreSitLoaded[i] = TextureCache.prefetch(OVERLAY_PRESIT_URL_ROOT + str + OVERLAY_PRESIT_URL_POSTFIX);
+                }
+                OVERLAY_PRESIT_URL_TEXT = TextureCache.prefetch(OVERLAY_PRESIT_URL_TEXT);
```
This is bad form, in our coding convention standard a variable with ALL_CAPS is assumed to be a constant.  but here are you are changing it.  I would use a new varaible named overlayPresetTexture or something instead instead of changing the value of OVERLAY_PRESIT_URL_TEXT.

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+                    var timePassed = now - overlayPreSitStart;
+                    if (timePassed >= OVERLAY_PRESIT_FRAME_DURATION) {
+                        overlayPreSitStart = now;
+                        index = index + 1;
```
Please give this variable a more descriptive name, I'm concerned that someone might inadvertently change this, because we use index as a variable in several other places in this file.  How about preSitLoadIndex?

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+
+            if (isInEditMode()) {
+                return;
+            }
```
Is returning what we want to do here?  I thought maybe we should hide the overlay instead?

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+        checkSeatForAvatar: function () {
+            return deadlockGate;
+
+            // var isOccupied = false;
```
Please remove this dead code.

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+            // get chairProperties to calculate middle of chair
+            seatCenterPosition = chairProperties.position;
+
+            var yOffset = chairProperties.dimensions.y / 2 * chairOffsetRatio;
```
This chairOffsetRatio maybe should be multiplied by some factor depending on the MyAvatar.getHeight().  so that tall avatars will have a large offset and short avatar would have a smaller offset.  For example:
```
var DEFAULT_AVATAR_HEIGHT = 1.755;
var chairOffsetRatio = 0.2 * (MyAvatar.getHeight() / DEFAULT_AVATAR_HEIGHT);
```

robin-k-wilson replied:
Flagging this comment to come back to.

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+        Settings.setValue(SETTING_KEY, entityID);
+        sitDownSettlePeriod = Date.now() + IK_SETTLE_TIME;
+
+        print(4);
```
please remove these debug statements.  Or add a global variable SITTING_DEBUG and do something like:
```
if (SITTING_DEBUG) {
    print("got here 4");
}
```

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+
+        print(4);
+
+        // MyAvatar.resetSensorsAndBody();
```
remove

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+
+        // MyAvatar.resetSensorsAndBody();
+
+        print(5);
```
remove

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+        MyAvatar.centerBody();
+        var index = MyAvatar.getJointIndex("Hips");
+
+        // var hipLocalRotation = MyAvatar.getJointRotation(index);
```
remove this dead code.

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+
+        Script.setTimeout(function () {
+
+            print(6, JSON.stringify(chairProperties.rotation));
```
remove

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+
+        }, SIT_DELAY);
+
+        animStateHandlerID = MyAvatar.addAnimationStateHandler(function (avatarProperties) {
```
Why is this here? We should not be directly changing how the IK works.  please remove.

robin-k-wilson replied:
Flagging for question on this function.

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+    }
+
+    function sitDown() {
+        print("SIT DOWN");
```
remove

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+    function sitDown() {
+        print("SIT DOWN");
+
+        // if (utils.checkSeatForAvatar()) {
```
remove

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+            if (avatarDistance < STANDUP_DISTANCE) {
+                var index = MyAvatar.getJointIndex("Hips");
+                MyAvatar.clearPinOnJoint(index);
+                var offset = { x: 0, y: 1.0, z: -1 / 2 - chairProperties.dimensions.z * chairProperties.registrationPoint.z };
```
the -1/2 should be replaced by a constant. like CHAIR_DISMOUNT_OFFSET or something.

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+
+        // Is the seat used
+        checkSeatForAvatar: function () {
+            return deadlockGate;
```
I'm confused is this used to check if the MyAvatar is in the chair? or if ANY avatar is sitting in the chair.

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+
+        utils.setHeadToHipsDistance();
+
+        lockChairOnStandUp = Entities.getEntityProperties(entityID, 'locked').locked;
```
Why do we need to lock/unlock the chair, it doesn't look like we are modifying it at all.

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+        location.hostChanged.disconnect(standUp);
+
+        canStand = false;
+        MyAvatar.removeAnimationStateHandler(animStateHandlerID);
```
remove

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+
+    // User can also click on overlay to sit down
+    this.mouseReleaseOnOverlay = function (overlayID, pointerEvent) {
+        print("Mouse release");
```
remove print

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+    this.mouseReleaseOnOverlay = function (overlayID, pointerEvent) {
+        print("Mouse release");
+        if (overlayID === overlaySittable && pointerEvent.isLeftButton) {
+            print("CAN SIT", deadlockGate, Settings.getValue(SETTING_KEY));
```
remove print

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+            if (HMD.active) {
+
+                var headWorldPosition;
+                if (MyAvatar.getJointIndex("Head") !== -1) {
```
I don't think we need to check if the avatar has a head here.  If we are in HMD mode we are guaranteed that getPoseValue(Head) will return something.

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+                } else {
+
+                    // Avatar has no head joint
+                    print("I haz no head!!!");
```
remove print

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+                }
+
+                var headDeviation = Vec3.distance(headWorldPosition, seatCenterPosition);
+                // print("headDeviation is ", headDeviation);
```
remove dead prints

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+
+            if (hasAvatarSpineError || hasHeldDriveKey || hasAvatarMovedTooFar) {
+                standUp();
+                print("ROBIN ", hasAvatarSpineError, hasHeldDriveKey, hasAvatarMovedTooFar);
```
remove print

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,692 @@
+                if ((deviationRatio > ONE_HUNDRED_AND_TEN_PERCENT || deviationRatio < EIGHTY_PERCENT) && 
+                    sitDownSettlePeriod && now > sitDownSettlePeriod) {
+
+                    // print("large deviation");
```
remove print

robin-k-wilson commented:
- Fixed linter errors and removed prints
- Changed variable names as mentioned by hyperlogic
- Added sitServer.js script to ensure only one Avatar can sit at a time
- Overlay updates depending on HMD vs Desktop mode

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,691 @@
+            Script.update.connect(update);
+            MyAvatar.scaleChanged.connect(standUp);
+            location.hostChanged.connect(standUp);
+            Script.scriptEnding.connect(standUp);
```
I think we also want to standUp if the MyAvatar.onLoadComplete signal is fired

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,691 @@
+        Script.update.disconnect(update);
+        MyAvatar.scaleChanged.disconnect(standUp);
+        location.hostChanged.disconnect(standUp);
+        Script.scriptEnding.disconnect(standUp);
```
And here disconnect from MyAvatar.onLoadComplete

hyperlogic commented on Prototyping/Sit/sit.js:
```diff
@@ -0,0 +1,691 @@
+                }
+            }
+
+            if (SITTING_DEBUG){
```
formatting

AndySeattle commented:
@robin-k-wilson I tried to download the JSON in the test plan above, and could not, i get a "failed - Forbidden" meaning access denied to the file. is it possible because this file was added before i was part of the group?

AndySeattle commented:
Howdy, @robin-k-wilson still seeing the same "failed"forbidden" error when I try to download the JSON file. This is blocking me from getting started, so going to mark this as QA failed to bubble up a bit :) 

![image](https://user-images.githubusercontent.com/41308582/43966584-221be020-9c77-11e8-9382-23f5a506e378.png)

AndySeattle commented:
Have a prototype of this running in my office space now, testing it out.

misslivirose commented:
@robin-k-wilson @AndySeattle are these the versions we're using actively right now? Can we get this PR finished up and merged?

AndySeattle commented:
We just reviewed and this and looks good, QA approved.

misslivirose commented:
Thank you!
